tools/nodejs/src/connection.cpp

```diff
@@ -66,6 +70,256 @@ Napi::Value Connection::Prepare(const Napi::CallbackInfo &info) {
+			auto return_string_vec_ptr = duckdb::FlatVector::GetData<duckdb::string_t>(*jsargs->result);
+
+			for (duckdb::idx_t row_idx = 0; row_idx < jsargs->rows; row_idx++) {
+				if (!return_validity[row_idx]) {
+					duckdb::FlatVector::SetNull(*jsargs->result, row_idx, true);
+				} else {
+					auto str = return_string_array.Get(row_idx).As<Napi::String>();
+					return_string_vec_ptr[row_idx] = duckdb::StringVector::AddString(*jsargs->result, str);
+				}
+			}
+		}
+	} catch (const std::exception &e) {
+		jsargs->error = e.what();
+	}
+	jsargs->done = true;
+}
+
+struct RegisterTask : public Task {
+	RegisterTask(Connection &connection_, std::string name_, std::string return_type_name_, Napi::Function callback_)
+	    : Task(connection_, callback_), name(name_), return_type_name(return_type_name_) {
+	}
+
+	void DoWork() override {
+		auto &connection = Get<Connection>();
+		auto &udf_ptr = connection.udfs[name];
+		duckdb::scalar_function_t udf_function = [&udf_ptr](duckdb::DataChunk &args, duckdb::ExpressionState &state,
+		                                                    duckdb::Vector &result) -> void {
+			// here we can do only DuckDB stuff because we do not have a functioning env
+
+			// Flatten all args to simplify udfs
+			args.Normalify();
+
+			JSArgs jsargs;
+			jsargs.rows = args.size();
+			jsargs.args = &args;
+			jsargs.result = &result;
+			jsargs.done = false;
+
+			udf_ptr.BlockingCall(&jsargs);
+			while (!jsargs.done) {
+				std::this_thread::yield();
+			}
+			if (!jsargs.error.empty()) {
+				throw duckdb::IOException(jsargs.error);
+			}
+		};
+
+		auto expr = duckdb::Parser::ParseExpressionList(duckdb::StringUtil::Format("asdf::%s", return_type_name));
+		auto &cast = (duckdb::CastExpression &)*expr[0];
+		auto return_type = cast.cast_type;
+
+		connection.connection->CreateVectorizedFunction(name, std::vector<duckdb::LogicalType> {}, return_type,
+		                                                udf_function, duckdb::LogicalType::ANY);
+	}
+	std::string name;
+	std::string return_type_name;
+};
+
+Napi::Value Connection::Register(const Napi::CallbackInfo &info) {
+	auto env = info.Env();
+	if (info.Length() < 3 || !info[0].IsString() || !info[1].IsString() || !info[2].IsFunction()) {
+		Napi::TypeError::New(env, "Holding it wrong").ThrowAsJavaScriptException();
+		return env.Null();
+	}
+
+	std::string name = info[0].As<Napi::String>();
+	std::string return_type_name = info[1].As<Napi::String>();
+	Napi::Function udf_callback = info[2].As<Napi::Function>();
+	Napi::Function completion_callback;
+	if (info.Length() > 3 && info[3].IsFunction()) {
+		completion_callback = info[3].As<Napi::Function>();
+	}
+
+	if (udfs.find(name) != udfs.end()) {
+		Napi::TypeError::New(env, "UDF with this name already exists").ThrowAsJavaScriptException();
+		return env.Null();
+	}
+
+	auto udf = DuckDBNodeUDFFUnction::New(env, udf_callback, "duckdb_node_udf" + name, 0, 1, nullptr,
```

Comment:
I think this 1 is connected to the thread limit of 1 in tools/nodejs/src/database.cpp:32?
Maybe we should make a constexpr of the thread limit and use it in both places.